enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -2822,6 +2825,442 @@ class AdjointGenerator
                        << I;
         report_fatal_error("(reverse) unknown intrinsic");
       }
+      return;
+    }
+    case DerivativeMode::ForwardMode: {
+
+      IRBuilder<> Builder2(&I);
+      getForwardBuilder(Builder2);
+
+      switch (ID) {
+      case Intrinsic::nvvm_barrier0_popc:
+      case Intrinsic::nvvm_barrier0_and:
+      case Intrinsic::nvvm_barrier0_or:
+      case Intrinsic::nvvm_barrier0:
+      case Intrinsic::amdgcn_s_barrier:
+      case Intrinsic::nvvm_membar_cta:
+      case Intrinsic::nvvm_membar_gl:
+      case Intrinsic::nvvm_membar_sys:
+      case Intrinsic::assume:
+      case Intrinsic::prefetch:
+      case Intrinsic::dbg_declare:
+      case Intrinsic::dbg_value:
+#if LLVM_VERSION_MAJOR > 6
+      case Intrinsic::dbg_label:
+#endif
+      case Intrinsic::dbg_addr:
+      case Intrinsic::floor:
+      case Intrinsic::ceil:
+      case Intrinsic::trunc:
+      case Intrinsic::rint:
+      case Intrinsic::nearbyint:
+      case Intrinsic::round:
+        // Derivative of these is zero and requires no modification
+        return;
+
+#if LLVM_VERSION_MAJOR >= 9
+#if LLVM_VERSION_MAJOR >= 12
+      case Intrinsic::vector_reduce_fadd:
+#else
+      case Intrinsic::experimental_vector_reduce_v2_fadd:
+#endif
+      {
+        if (gutils->isConstantInstruction(&I))
+          return;
+
+        if (!gutils->isConstantValue(orig_ops[1])) {
+          auto und = UndefValue::get(orig_ops[1]->getType());
+          auto mask = ConstantAggregateZero::get(VectorType::get(
+              Type::getInt32Ty(und->getContext()),
+#if LLVM_VERSION_MAJOR >= 11
+              cast<VectorType>(und->getType())->getElementCount()));
+#else
+              cast<VectorType>(und->getType())->getNumElements()));
+#endif
+          auto accdiff = diffe(orig_ops[0], Builder2);
+          auto diff = Builder2.CreateShuffleVector(
```

Comment:
This is a reverse mode derivative, in forward mode I think the correct answer is to sum the shadows.